### ISSUE TITLE
GraalVM should convert signals from native code to exceptions so that JVM don't crash

### ISSUE THREAD

**Issue by preetnitkkr** [oracle/graal#602]
Hi,

Consider a situation in which native code throw some signal e.g. SIGSEGV in certain situations. In these cases, the entire JVM shouldn't crash in a production environment. 
I tried with a sample code and JVM crashes. The code is as follow -
// hello.c
```
#include "signal.h"
int main() 
{      raise(SIGSEGV); 	return 0;  }
```

// Crash.java
```
import java.io.*
import org.graalvm.polyglot.*;
public class Crash {
	public static void main(String[] args) {
		try (Context context = Context.newBuilder().allowNativeAccess(true).build()) {
			for(int i = 0;i<5;i++) {
				final Value library;
				try {
					library = context.eval(Source.newBuilder("llvm", new File("hello.bc")).build());
					final Value main = library.getMember("main");
					main.execute();
				} catch (Exception e) {
					e.printStackTrace();
					return;
				}
			}
		}
	}
}
```

Is there any method to stop the JVM crash ?

Thanks !

**Comment by preetnitkkr** [comment]
The error seen after JVM crashes is:
```
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x00007f644cd58428, pid=24667, tid=0x00007f644d928700
#
# JRE version: Java(TM) SE Runtime Environment (8.0_172-b11) (build 1.8.0_172-b11)
# Java VM: GraalVM 1.0.0-rc4 (25.71-b01-internal-jvmci-0.45 mixed mode linux-amd64 compressed oops)
# Problematic frame:
# C  [libc.so.6+0x35428]  gsignal+0x38
#
# Failed to write core dump. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
#
# An error report file with more information is saved as:
# /home/preet/Adobe/demos/crash/hs_err_pid24667.log
#
# If you would like to submit a bug report, please visit:
#   http://bugreport.java.com/bugreport/crash.jsp
# The crash happened outside the Java Virtual Machine in native code.
# See problematic frame for where to report the bug.
#
Aborted (core dumped)
```

**Comment by rschatz** [comment]
I'm assuming you're asking about implicit segfaults, for example because of illegal memory access.

Our llvm runtime does not hook into the JVM’s signal handling. Handling errors at that level is close to impossible, since the system is most likely already in a corrupted state. So this just delays the problem until later, and doesn’t add to overall stability.

On the other hand, if you literally meant running `raise(SIGSEGV)` as a way of non-local return, you might want to consider using C++ exceptions instead. They are catchable from Java.

**Comment by preetnitkkr** [comment]
> I'm assuming you're asking about implicit segfaults, for example because of illegal memory access.
yes, implicit segfaults, crash due to NULL pointer dereference, dangling pointer etc. - all types of memory safety related problems.

I guess this issue is getting related to https://github.com/oracle/graal/issues/605 . 
So closing this as it has become duplicate of that.

Thanks for the help !

**Comment by rschatz** [comment]
We have just published an experimental option for making LLVM on GraalVM memory-safe. See https://medium.com/graalvm/safe-and-sandboxed-execution-of-native-code-f6096b35c360 for more details.
